### DISCUSSION TITLE
Doge Community Tooling: Dogecoin Client Libraries, Block Explorer and Scripting IDE

### DISCUSSION THREAD

**Discussion by cf** [dogecoin/dogecoin#3566] [Show and tell]
Hello all! 
Over the past few months of working in the doge ecosystem, we have built and recently open sourced some tooling for doge, and I wanted to share here in case it would helpful to anyone else/anyone wanted to help contribute!

## [doge-sdk](https://github.com/QEDProtocol/doge-sdk/)
<a href="https://github.com/QEDProtocol/doge-sdk/"><img src="https://fs.qed.zone/imgs/doge-sdk-screenshot.png" alt="Doge SDK Screenshot" width="400" /></a>

An ultra lightweight (20kb gzipped) doge library for Node.js and the Browser.
* Only one pinned dependency, [@noble/secp256k1](https://github.com/paulmillr/noble-secp256k1), which has already received a  [comprehensive security audit](https://cure53.de/pentest-report_noble-lib.pdf).
  * 🐮 Bullwork against supply chain attacks
* Supports out of the box P2PKH + P2SH transactions
* Comes out of the box with a Memory wallet 
* Easy interface to support hardware wallets, and includes a reference implementation for [Ledger support](https://github.com/QEDProtocol/doge-sdk/blob/main/docs/ThirdPartyWallets.md#ledger-reference-implementation)
* Full RPC Client with Types + Doge Electrs Client for production environments where you don't want to expose the standard dogecoin rpc API
* Supports custom HTTP transports with a default fallback using the [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)
* Includes a compiler/assembler/disassembler for BASM-lite + standard bitcoin VM assembly
* Does not depend on bitcoinlib-js =D
* Easy to build multi-platform wallets and dogecoin apps

## [electrs-doge](https://github.com/QEDProtocol/electrs-doge)

<a href="https://doge-explorer.qed.me"><img src="https://fs.qed.zone/imgs/doge-explorer-screenshot.png" alt="Doge Electrs Screenshot" width="400" /></a>

A production-ready full history API for building block explorers and indexing dogecoin transactions.
Checkout an example block explorer built with electrs-doge: https://doge-explorer.qed.me/
* API fully compatible with [Blockstream's bitcoin electrs implementation](https://github.com/Blockstream/electrs).
* Works out of the box with Blockstream's esplora block explorer, check out https://doge-explorer.qed.me/ for an example implementation
* Supports indexing of AuxPow blocks/other dogecoin specific quirks
* Super fast, stable and proven in production

## BitIDE Doge
<img src="https://fs.qed.zone/imgs/doge-bitide-screenshot.png" alt="Doge BitIDE Screenshot" width="400" />

A fully featured Dogecoin scripting IDE with bundled local regtest testnet, scripting IDE, block explorer and electrs API.

One-liner installation + startup (no dependencies other than docker):
```bash
docker run -p 1337:1337 -it --rm qedprotocol/bitide-doge:latest
```
IDE: http://localhost:1337
Block Explorer: http://localhost:1337/explorer/
Electrs API: http://localhost:1337/api/
RPC API: http://devnet:devnet@localhost:1337/bitcoin-rpc/

### Features
* Full P2SH scripting IDE with autocomplete, VM emulator, stack explorer and extended scripting language (BASM)
* Full Web GUI/REPL for dogecoin RPC commands
* Easy to use Web P2SH GUI for funding and spending scripts
* Web GUI Wallet Manager + Signature Signing helpers for spending scripts which include signature checks

## [rust-dogecoin](https://github.com/QEDProtocol/rust-dogecoin)
A work in progress dogecoin compatible fork of rust-bitcoin that is already being used by electrs-doge (please submit a PR if you find any functionality that is missing or 
* Supports parsing of dogecoin block files, graceful serialization and deserialization of AuxPow + non-AuxPow blocks
* Supports dogecoin wallets, addresses and other functions for dogecoin apps in rust

If anyone is interested in contributing or wants to work on other community tooling ping us on X ([@cmpeq](https://x.com/cmpeq) + [@QEDProtocol](https://x.com/QEDProtocol))

**Comment by patricklodder** [comment]
Very nice!

> ### electrs-doge

Would you mind enabling issues on this repo? I have questions about the [diff](https://github.com/Blockstream/electrs/compare/new-index...QEDProtocol:electrs-doge:doge).

**Reply by cf** [reply]
just realized I hadn't pushed the commit that removed some of the debug logging, should be more clear now =)

> Very nice!
> 
> > ### electrs-doge
> 
> Would you mind enabling issues on this repo? I have questions about the [diff](https://github.com/Blockstream/electrs/compare/new-index...QEDProtocol:electrs-doge:doge).

**Reply by patricklodder** [reply]
Thanks, now I have a lot less questions 😂

**Comment by patiee** [comment]
Great job!

**Reply by Neotravolta** [reply]
Keep buidling folks 🏁

**Comment by hieusats** [comment]
**Seeking QEDProtocol Dev Team Support for Updating Ord-Dogecoin and Expanding the Doginals Ecosystem
Dear QEDProtocol Development Team,**
We extend our heartfelt gratitude for your unwavering support and contributions to the Dogecoin ecosystem. We're excited to share that Doginals, a perfect fork of Ordinals, is now operational on the Dogecoin network. However, we are facing some challenges that require immediate attention and expertise.

The Ord-Dogecoin project, originally forked from the Ordinals repository by the developer of Apezord, has not been updated for over a year. This has resulted in significant issues with indexing DRC20 tokens, hampering the growth and functionality of the Doginals ecosystem.

**Call for QEDProtocol Development Team Support**
We urgently seek the support of the QEDProtocol development team to lead the effort in updating the Ord-Dogecoin repository to support the latest features and improvements. Our goal is to align Ord-Dogecoin with the latest version of Ordinals, ensuring full compatibility and enhanced performance for the Doginals ecosystem.

**Current Repository and Reference**
For reference, you can view the current state of the Ord-Dogecoin project here: [Apezord's Ord-Dogecoin Repository](https://github.com/apezord/ord-dogecoin).

**Next Steps**
**_Update the Repository:_** The immediate task is to update Ord-Dogecoin to match the latest version of Ordinals. This will resolve existing issues and enable new features.
**_Enhance Indexing:_** Address and improve the indexing capabilities for DRC20 tokens.
**_Community Collaboration:_** We welcome collaboration and contributions from developers and enthusiasts to foster the growth of Doginals.
**How You Can Help**
If you have experience with blockchain development and are passionate about the Dogecoin ecosystem, we encourage you to contribute to the Ord-Dogecoin project. Together, we can ensure that Doginals reaches its full potential.

For more details and to join the development effort, please visit our repository: [Ord-Dogecoin on GitHub](https://github.com/apezord/ord-dogecoin).

Thank you for your continued support. Let's work together to build a brighter future for Dogecoin and Doginals.

Sincerely,
The Doginals community
